File: predict-java-nextInt.java
/**
java版本：根据前两次nextInt()的值获取下一个nextInt的值
*/
import java.util.Random;
public class Demo01 {
    private static final long multiplier = 0x5DEECE66DL;
    private static final long addend = 0xBL;
    private static final long mask = (1L << 48) - 1;

    public static void main(String[] paramArrayOfString) {
        Random random = new Random();
        random.setSeed(System.currentTimeMillis());

        int a = random.nextInt();
        int b = random.nextInt();
        long oldseed = ((long) a << 16);
        for (int i = 0; i <= 0xffff; i++) {//暴力破解种子
            long nextseed = (oldseed * multiplier + addend) & mask;
            if ((int) (nextseed >>> 16) == b) {
                oldseed = nextseed;
                break;
            }
            oldseed += 1;
        }
        int next_int = (int) (((oldseed * multiplier + addend) & mask) >>> 16);

        System.out.println(next_int);
        System.out.println(random.nextInt());
    }
}

File: predict-java-nextInt.py
"""
Random random = new Random();
random.setSeed(System.currentTimeMillis());
for (int i = 1; i <= 3; ++i) {
    System.out.println(random.nextInt());
}
python版本：根据前两次nextInt()的值获取下一个nextInt的值
"""
sequence = [-2037540693, -1544783484]

multiplier = 0x5DEECE66D
addend = 0xB
mask = (1 << 48) - 1


def d(n):  # 补码处理
    if bin(abs(n))[2:].zfill(32)[0] == '1':
        n = ((-n) ^ (0xffffffff)) + 1
    return n


oldseed = sequence[0] << 16
for i in range(0xffff):  # 暴力破解种子
    nextseed = (oldseed * multiplier + addend) & mask
    if d(nextseed >> 16) == sequence[1]:
        oldseed = nextseed
        break
    oldseed += 1
next_int = d(((oldseed * multiplier + addend) & mask) >> 16)

print(next_int)
